### PR TITLE
[Contacts list] Send an invite by email to enroll a new user

### PR DESCRIPTION
**Pull request by yostyle**
#86

### CODE REVIEW

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListViewState.kt [hunk 1]
```diff
@@ -40,9 +41,18 @@ data class TchapContactListViewState(
...
+        private val showInviteActions: Boolean
 ) : MvRxState {
     constructor(args: TchapContactListFragmentArgs) : this(
-            excludedUserIds = args.excludedUserIds
+            excludedUserIds = args.excludedUserIds,
+            singleSelection = args.singleSelection,
+            showInviteActions = args.showInviteActions
     )
+
+    fun showInviteActions() = showInviteActions
```
Comment (Florian14) [inline vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListViewState.kt:57]: remove and set attribute public

File: vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt [hunk 2]
```diff
@@ -124,6 +137,120 @@ class HomeDetailViewModel @AssistedInject constructor(@Assisted initialState: Ho
...
+            if (data.isNullOrEmpty()) {
+                _viewEvents.post(TchapHomeViewEvents.GetPlatform(action.email))
+            } else {
+                _viewEvents.post(TchapHomeViewEvents.InviteIgnoredForDiscoveredUser(action.email))
+            }
+        }
+    }
+
+    private fun handleUnauthorizeEmail() = withState {
+        if (it.inviteEmail != null) {
```
Comment (Florian14) [inline vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt:163]: we can remove one imbrication level by returning from the function (it.inviteEmail ?: return@withState)

File: vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt [hunk 3]
```diff
@@ -124,6 +137,120 @@ class HomeDetailViewModel @AssistedInject constructor(@Assisted initialState: Ho
...
+                _viewEvents.post(TchapHomeViewEvents.InviteIgnoredForUnauthorizedEmail(it.inviteEmail))
+            } else {
+                // Ignore the error, notify the user that the invite has been already sent
+                _viewEvents.post(TchapHomeViewEvents.InviteIgnoredForExistingRoom(it.inviteEmail))
+            }
+        }
+    }
+
+    private fun handleCreateDiscussion(action: HomeDetailAction.CreateDiscussion) = withState {
+        if (it.inviteEmail != null) {
```
Comment (Florian14) [inline vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt:174]: idem

File: vector/src/main/java/im/vector/app/features/home/HomeDetailAction.kt [hunk 4]
```diff
@@ -16,9 +16,13 @@
...
 package im.vector.app.features.home
 
+import fr.gouv.tchap.android.sdk.internal.services.threepidplatformdiscover.model.Platform
 import im.vector.app.core.platform.VectorViewModelAction
 
 sealed class HomeDetailAction : VectorViewModelAction {
     data class SwitchDisplayMode(val displayMode: RoomListDisplayMode) : HomeDetailAction()
     object MarkAllRoomsRead : HomeDetailAction()
+    data class InviteByEmail(val email: String) : HomeDetailAction()
+    object UnauthorizeEmail : HomeDetailAction()
```
Comment (Florian14) [inline vector/src/main/java/im/vector/app/features/home/HomeDetailAction.kt:26]: typo : Unauthorize**d**Email

File: vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt [hunk 5]
```diff
@@ -124,6 +137,120 @@ class HomeDetailViewModel @AssistedInject constructor(@Assisted initialState: Ho
...
+
+            if (data.isNullOrEmpty()) {
+                _viewEvents.post(TchapHomeViewEvents.GetPlatform(action.email))
+            } else {
+                _viewEvents.post(TchapHomeViewEvents.InviteIgnoredForDiscoveredUser(action.email))
+            }
+        }
+    }
+
+    private fun handleUnauthorizeEmail() = withState {
```
Comment (Florian14) [inline vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt:162]: typo: handleUnauthorize**ed**Email

File: vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt [hunk 6]
```diff
@@ -124,6 +137,120 @@ class HomeDetailViewModel @AssistedInject constructor(@Assisted initialState: Ho
...
+            } else {
+                // Ignore the error, notify the user that the invite has been already sent
+                _viewEvents.post(TchapHomeViewEvents.InviteIgnoredForExistingRoom(it.inviteEmail))
+            }
+        }
+    }
+
+    private fun handleCreateDiscussion(action: HomeDetailAction.CreateDiscussion) = withState {
+        if (it.inviteEmail != null) {
+            // Pursue the invite process by checking whether an invite has been already sent
```
Comment (Florian14) [inline vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt:175]: Pursue -> Continue ?

File: vector/src/main/java/im/vector/app/features/home/HomeDetailFragment.kt [hunk 7]
```diff
@@ -194,12 +218,39 @@ class HomeDetailFragment @Inject constructor(
...
+                is PlatformViewEvents.Loading -> showLoading(it.message)
+                is PlatformViewEvents.Failure -> viewModel.handle(HomeDetailAction.UnauthorizeEmail)
+                is PlatformViewEvents.Success -> viewModel.handle(HomeDetailAction.CreateDiscussion(it.platform))
+            }.exhaustive
+        }
+
+        viewModel.observeViewEvents {
+            when (it) {
+                is TchapHomeViewEvents.InviteIgnoredForDiscoveredUser       -> {
+                    Toast.makeText(requireContext(), getString(R.string.tchap_discussion_already_exist, it.email), Toast.LENGTH_LONG).show()
```
Comment (giomfo) [inline vector/src/main/java/im/vector/app/features/home/HomeDetailFragment.kt:239]: TODO: open the DM for this discovered user #126

File: vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt [hunk 8]
```diff
@@ -124,6 +137,120 @@ class HomeDetailViewModel @AssistedInject constructor(@Assisted initialState: Ho
...
+                        createDiscussion(it.inviteEmail)
+                    }
+                } else {
+                    _viewEvents.post(TchapHomeViewEvents.InviteIgnoredForUnauthorizedEmail(it.inviteEmail))
+                }
+            } else {
+                // There is already a discussion with this email
+                // We do not re-invite the NoTchapUser except if
+                // the email is bound to the external instance (for which the invites may expire).
+                if (action.platform.hs.isNotEmpty()) {
```
Comment (giomfo) [inline vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt:189]: You forgot to check if the hs is the external one ;)

File: vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt [hunk 9]
```diff
@@ -124,6 +137,120 @@ class HomeDetailViewModel @AssistedInject constructor(@Assisted initialState: Ho
...
+                } else {
+                    // Notify the user that the invite has been already sent
+                    _viewEvents.post(TchapHomeViewEvents.InviteIgnoredForExistingRoom(it.inviteEmail))
+                }
+            }
+        }
+    }
+
+    private suspend fun createDiscussion(email: String) {
+        val adminE2EByDefault = rawService.getElementWellknown(session.myUserId)
```
Comment (giomfo) [inline vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt:210]: Please remove this adminE2EEDefault, I will explain in direct

File: vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt [hunk 10]
```diff
@@ -124,6 +137,120 @@ class HomeDetailViewModel @AssistedInject constructor(@Assisted initialState: Ho
...
+                ?.isE2EByDefault()
+                ?: true
+
+        val roomParams = CreateRoomParams()
+                .apply {
+                    invite3pids.add(ThreePid.Email(email))
+                    setDirectMessage()
+                    enableEncryptionIfInvitedUsersSupportIt = adminE2EByDefault
```
Comment (giomfo) [inline vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt:218]: we don't need to set this value `enableEncryptionIfInvitedUsersSupportIt`. We let the serveur force the encryption in DM

File: vector/src/main/java/im/vector/app/features/home/HomeDetailAction.kt [hunk 11]
```diff
@@ -16,9 +16,13 @@
...
+    data class CreateDiscussion(val platform: Platform) : HomeDetailAction()
```
Comment (giomfo) [inline vector/src/main/java/im/vector/app/features/home/HomeDetailAction.kt:27]: I would like to discuss with you about the param here

File: vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt (hunk 6 above)
Comment (yostyle) [reply vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt]: Same comment as element ios

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListViewModel.kt [hunk 12]
```diff
@@ -54,7 +56,8 @@ private typealias DirectoryUsersSearch = String
 
 class TchapContactListViewModel @AssistedInject constructor(@Assisted initialState: TchapContactListViewState,
                                                             private val contactsDataSource: ContactsDataSource,
-                                                            private val session: Session)
+                                                            private val session: Session,
+                                                            private val matrix: Matrix)
```
[the remark this answers is not in the document]
Comment (yostyle) [reply vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListViewModel.kt]: I don't need Matrix instance here. I forgot to remove it.

File: vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt [hunk 13]
```diff
@@ -124,6 +135,108 @@ class HomeDetailViewModel @AssistedInject constructor(@Assisted initialState: Ho
...
+            // Ignore the error, notify the user that the invite has been already sent
+            _viewEvents.post(TchapHomeViewEvents.InviteIgnoredForExistingRoom(it.inviteEmail))
+        }
+    }
+
+    private fun handleCreateDiscussion(action: HomeDetailAction.CreateDiscussion) = withState {
+        it.inviteEmail ?: return@withState
+
+        if (it.existingRoom.isNullOrEmpty()) {
+            // Send the invite if the email is authorized
```
Comment (giomfo) [inline vector/src/main/java/im/vector/app/features/home/HomeDetailViewModel.kt:175]: this comment may be removed

**Review by giomfo — APPROVED** [review]
just one comment to removed

**Pull request merged**